File: DebateFeedback/Features/DebateTimer/TimerViewModel.swift
//
//  TimerViewModel.swift
//  DebateFeedback
//
//

import Foundation
import SwiftData
import AVFoundation

@Observable
@MainActor
final class TimerViewModel {
    let debateSession: DebateSession
    private let modelContext: ModelContext

    // Services
    private var timerService: TimerService
    private let audioService = AudioRecordingService()
    private let playbackService = AudioPlaybackService()
    private let uploadService = UploadService.shared

    // State
    private(set) var currentSpeakerIndex = 0
    private(set) var speakers: [(name: String, position: String, studentId: String)] = []
    private(set) var isRecording = false
    private(set) var currentRecordingURL: URL?
    private(set) var recordings: [SpeechRecording] = []

    // Playback State
    private(set) var playingRecordingId: UUID?
    var isPlaying: Bool {
        playbackService.isPlaying
    }

    // UI State
    var showError = false
    var errorMessage = ""
    var uploadProgress: [UUID: Double] = [:]

    // Warning state tracking (to avoid repeated haptics)
    @ObservationIgnored private var has60sWarningFired = false
    @ObservationIgnored private var has30sWarningFired = false
    @ObservationIgnored private var has15sWarningFired = false

    init(debateSession: DebateSession, modelContext: ModelContext) {
        self.debateSession = debateSession
        self.modelContext = modelContext
        self.timerService = TimerService(speechDuration: TimeInterval(debateSession.speechTimeSeconds))

        setupSpeakers()
        loadExistingRecordings()
        checkMicrophonePermission()
    }

    private func loadExistingRecordings() {
        if let existingRecordings = debateSession.speechRecordings {
            let standardArray: [SpeechRecording] = existingRecordings
            self.recordings = standardArray.sorted { $0.createdAt < $1.createdAt }
        }
    }

    // MARK: - Setup

    private func setupSpeakers() {
        guard let composition = debateSession.teamComposition else { return }
        let speakerOrder = composition.getSpeakerOrder(format: debateSession.format)

        // Convert to the format expected by speakers array
        speakers = speakerOrder.map { (studentId, position) in
            // Find student name from debateSession.students
            let studentName = debateSession.students?.first(where: { $0.id.uuidString == studentId })?.name ?? "Unknown"
            return (name: studentName, position: position, studentId: studentId)
        }
    }

    private func checkMicrophonePermission() {
        Task {
            let granted = await audioService.requestPermission()
            if !granted {
                errorMessage = Constants.ErrorMessages.microphonePermissionDenied
                showError = true
            }
        }
    }

    // MARK: - Timer Control

    var timerState: TimerService.State {
        timerService.state
    }

    var elapsedTime: TimeInterval {
        timerService.elapsedTime
    }

    var formattedTime: String {
        timerService.formattedTime
    }

    var isOvertime: Bool {
        timerService.isOvertime
    }

    var progressPercentage: Double {
        timerService.progressPercentage
    }

    // MARK: - Smart Timer Features

    /// Time remaining until expected end
    var timeRemaining: TimeInterval {
        let expectedDuration = TimeInterval(debateSession.speechTimeSeconds)
        return max(0, expectedDuration - elapsedTime)
    }

    /// Warning state based on time remaining
    enum WarningLevel {
        case none
        case oneMinute      // 60 seconds remaining
        case thirtySeconds  // 30 seconds remaining
        case fifteenSeconds // 15 seconds remaining
    }

    var currentWarningLevel: WarningLevel {
        if isOvertime {
            return .none
        }

        let remaining = timeRemaining
        if remaining <= 15 {
            return .fifteenSeconds
        } else if remaining <= 30 {
            return .thirtySeconds
        } else if remaining <= 60 {
            return .oneMinute
        } else {
            return .none
        }
    }

    /// Check and fire warnings with haptic feedback
    func checkAndFireWarnings() {
        guard isRecording && !isOvertime else {
            // Reset warning flags when not recording or in overtime
            if !isRecording {
                has60sWarningFired = false
                has30sWarningFired = false
                has15sWarningFired = false
            }
            return
        }

        switch currentWarningLevel {
        case .oneMinute:
            if !has60sWarningFired {
                HapticManager.shared.warning()
                has60sWarningFired = true
            }
        case .thirtySeconds:
            if !has30sWarningFired {
                HapticManager.shared.warning()
                has30sWarningFired = true
            }
        case .fifteenSeconds:
            if !has15sWarningFired {
                HapticManager.shared.warning()
                has15sWarningFired = true
            }
        case .none:
            break
        }
    }

    func startTimer() {
        guard !isRecording else { return }

        do {
            // Start recording
            let url = try audioService.startRecording(
                debateId: debateSession.id.uuidString,
                speakerName: currentSpeaker.name,
                position: currentSpeaker.position
            )

            currentRecordingURL = url
            isRecording = true

            // Start timer
            timerService.start()

        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    func stopTimer() {
        guard isRecording else { return }

        // Stop recording
        guard let result = audioService.stopRecording() else {
            // If recording failed or was already stopped, force reset state
            isRecording = false
            currentRecordingURL = nil
            timerService.stop()
            
            errorMessage = "Recording failed or was stopped unexpectedly"
            showError = true
            return
        }

        let minimumDuration: TimeInterval = 1.0
        if result.duration < minimumDuration {
            try? FileManager.deleteAudioFile(at: result.url.path)
            errorMessage = "Recording is too short. Please record at least one second of audio."
            showError = true
            return
        }

        // Stop timer
        timerService.stop()

        isRecording = false
        currentRecordingURL = nil

        // Create recording record
        let recording = SpeechRecording(
            speakerName: currentSpeaker.name,
            speakerPosition: currentSpeaker.position,
            localFilePath: result.url.path,
            durationSeconds: Int(result.duration),
            debateSession: debateSession
        )

        modelContext.insert(recording)
        recordings.append(recording)

        try? modelContext.save()

        // Start upload
        uploadRecording(recording)

        // Move to next speaker automatically
        if currentSpeakerIndex < speakers.count - 1 {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
                self?.nextSpeaker()
            }
        }
    }

    /// Ring the bell manually when user taps the bell icon
    func ringBell() {
        timerService.ringBellManually()
    }

    // MARK: - Speaker Navigation

    var currentSpeaker: (name: String, position: String, studentId: String) {
        speakers[currentSpeakerIndex]
    }

    var canGoBack: Bool {
        currentSpeakerIndex > 0
    }

    var canGoForward: Bool {
        currentSpeakerIndex < speakers.count - 1
    }

    var speakerProgress: String {
        "\(currentSpeakerIndex + 1) / \(speakers.count)"
    }

    func nextSpeaker() {
        guard canGoForward && !isRecording else { return }

        currentSpeakerIndex += 1
        timerService.reset()

        // Reset warning flags for new speaker
        has60sWarningFired = false
        has30sWarningFired = false
        has15sWarningFired = false
    }

    func previousSpeaker() {
        guard canGoBack && !isRecording else { return }

        currentSpeakerIndex -= 1
        timerService.reset()

        // Reset warning flags for new speaker
        has60sWarningFired = false
        has30sWarningFired = false
        has15sWarningFired = false
    }

    // MARK: - Upload Management

    private func uploadRecording(_ recording: SpeechRecording) {
        Task {
            do {
                recording.uploadStatus = .uploading
                recording.transcriptionStatus = .processing
                recording.feedbackStatus = .pending
                recording.updateAggregatedStatus()
                try? modelContext.save()

                let speechId = try await uploadService.uploadSpeech(
                    recording: recording,
                    debateSession: debateSession
                ) { progress in
                    Task { @MainActor in
                        recording.uploadProgress = progress
                        self.uploadProgress[recording.id] = progress
                    }
                }

                recording.uploadStatus = .uploaded
                recording.feedbackStatus = .processing
                recording.updateAggregatedStatus()
                recording.speechId = speechId // Store the speech ID
                try? modelContext.save()

                // Start polling for feedback
                pollForFeedback(recording: recording, speechId: speechId)

            } catch {
                recording.uploadStatus = .failed
                recording.transcriptionStatus = .failed
                recording.transcriptionErrorMessage = "Upload failed: \(error.localizedDescription)"
                recording.updateAggregatedStatus()
                try? modelContext.save()

                errorMessage = "Failed to upload \(recording.speakerName)'s speech: \(error.localizedDescription)"
                showError = true
            }
        }
    }

    func retryUpload(for recording: SpeechRecording) {
        uploadRecording(recording)
    }

    // MARK: - Feedback Polling

    private func pollForFeedback(recording: SpeechRecording, speechId: String) {
        Task {
            var attempts = 0
            let maxAttempts = 60 // 5 minutes max

            while attempts < maxAttempts {
                try? await Task.sleep(nanoseconds: UInt64(Constants.API.feedbackPollingInterval * 1_000_000_000))

                do {
                    let status: SpeechStatusResponse = try await APIClient.shared.request(
                        endpoint: .getSpeechStatus(speechId: speechId)
                    )

                    let isTerminal = apply(statusResponse: status, to: recording)
                    if isTerminal {
                        return
                    }

                } catch {
                    print("Polling error: \(error)")
                }

                attempts += 1
            }

            // Timeout
            recording.feedbackStatus = .failed
            recording.feedbackErrorMessage = "Timed out waiting for feedback"
            recording.updateAggregatedStatus()
            try? modelContext.save()
        }
    }

    private func apply(statusResponse: SpeechStatusResponse, to recording: SpeechRecording) -> Bool {
        let lowerStatus = statusResponse.status.lowercased()

        if let transStatus = statusResponse.transcriptionStatus {
            recording.transcriptionStatus = ProcessingStatus(apiStatus: transStatus)
        } else if lowerStatus == "complete" && recording.transcriptionStatus != .complete {
            recording.transcriptionStatus = .complete
        }

        if let transError = statusResponse.transcriptionError, !transError.isEmpty {
            recording.transcriptionErrorMessage = transError
            recording.transcriptionStatus = .failed
        }

        if let feedbackStatus = statusResponse.feedbackStatus {
            recording.feedbackStatus = ProcessingStatus(apiStatus: feedbackStatus)
        } else if lowerStatus == "complete" {
            recording.feedbackStatus = .complete
        }

        if let feedbackError = statusResponse.feedbackError, !feedbackError.isEmpty {
            recording.feedbackErrorMessage = feedbackError
            recording.feedbackStatus = .failed
        }

        if let docUrl = statusResponse.googleDocUrl, !docUrl.isEmpty {
            recording.feedbackUrl = docUrl
        }

        if let transcriptUrl = statusResponse.transcriptUrl ?? statusResponse.transcriptDownloadUrl, !transcriptUrl.isEmpty {
            recording.transcriptUrl = transcriptUrl
        }

        if let transcriptText = statusResponse.transcriptText, !transcriptText.isEmpty {
            recording.transcriptText = transcriptText
        }

        if let generalError = statusResponse.errorMessage, !generalError.isEmpty {
            if recording.feedbackStatus == .failed {
                recording.feedbackErrorMessage = generalError
            } else if recording.transcriptionStatus == .failed {
                recording.transcriptionErrorMessage = generalError
            }
        }

        recording.updateAggregatedStatus()
        try? modelContext.save()

        if recording.transcriptionStatus == .failed || recording.feedbackStatus == .failed {
            return true
        }

        if recording.feedbackStatus == .complete {
            return true
        }

        return false
    }

    // MARK: - Completion

    var isDebateComplete: Bool {
        recordings.count == speakers.count
    }

    var completedSpeeches: Int {
        recordings.count
    }

    func viewFeedback() {
        // Navigation handled by coordinator
    }

    // MARK: - Playback Management

    func togglePlayback(for recording: SpeechRecording) {
        // If this recording is currently playing, pause it
        if playingRecordingId == recording.id && isPlaying {
            playbackService.pause()
            return
        }

        // If this recording was paused, resume it
        if playingRecordingId == recording.id && !isPlaying {
            playbackService.resume()
            return
        }

        // Otherwise, start playing this recording
        playRecording(recording)
    }

    private func playRecording(_ recording: SpeechRecording) {
        let url = URL(fileURLWithPath: recording.localFilePath)

        // Check if file exists
        guard FileManager.default.fileExists(atPath: recording.localFilePath) else {
            errorMessage = "Recording file not found. It may have been deleted."
            showError = true
            return
        }

        do {
            try playbackService.play(from: url)
            playingRecordingId = recording.id
        } catch {
            errorMessage = "Failed to play recording: \(error.localizedDescription)"
            showError = true
        }
    }

    func stopPlayback() {
        playbackService.stop()
        playingRecordingId = nil
    }

    func getPlaybackProgress(for recording: SpeechRecording) -> Double {
        guard playingRecordingId == recording.id else { return 0 }
        return playbackService.progress
    }

    func getPlaybackTime(for recording: SpeechRecording) -> String {
        guard playingRecordingId == recording.id else {
            return formatDuration(recording.durationSeconds)
        }
        return "\(playbackService.formattedCurrentTime) / \(playbackService.formattedDuration)"
    }

    private func formatDuration(_ seconds: Int) -> String {
        let minutes = seconds / 60
        let secs = seconds % 60
        return String(format: "%d:%02d", minutes, secs)
    }

    // MARK: - Cleanup

    deinit {
        // Note: Cannot reliably cancel recording in deinit due to async/MainActor requirements
        // Recording will be stopped when the service is deallocated
        playbackService.stop()
    }
}

File: android/app/src/main/java/com/debatefeedback/ui/timer/TimerViewModel.kt
package com.debatefeedback.ui.timer

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.debatefeedback.core.Constants
import com.debatefeedback.data.remote.api.SpeechStatusResponse
import com.debatefeedback.domain.model.DebateSession
import com.debatefeedback.domain.model.ProcessingStatus
import com.debatefeedback.domain.model.SpeechRecording
import com.debatefeedback.domain.model.TeamComposition
import com.debatefeedback.domain.model.UploadStatus
import com.debatefeedback.repository.DebateRepository
import com.debatefeedback.repository.UploadRepository
import com.debatefeedback.services.AudioRecordingService
import com.debatefeedback.services.TimerService
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import java.io.File

class TimerViewModel(
    private val sessionId: String,
    private val debateRepository: DebateRepository,
    private val uploadRepository: UploadRepository,
    private val recordingService: AudioRecordingService,
    private val timerServiceFactory: (Int) -> TimerService
) : ViewModel() {

    private val _state = MutableStateFlow(TimerUiState())
    val state: StateFlow<TimerUiState> = _state

    private var timerService: TimerService? = null

    init {
        viewModelScope.launch { loadSession() }
        viewModelScope.launch {
            debateRepository.observeRecordings(sessionId).collect { recordings ->
                _state.value = _state.value.copy(recordings = recordings)
            }
        }
    }

    private suspend fun loadSession() {
        val session = debateRepository.getSession(sessionId)
        if (session == null) {
            _state.value = _state.value.copy(isLoading = false, errorMessage = "Session not found")
            return
        }
        timerService = timerServiceFactory(session.speechTimeSeconds).also { service ->
            service.onBell { count ->
                _state.value = _state.value.copy(lastBellCount = count, bellTimestamp = System.currentTimeMillis())
            }
            viewModelScope.launch {
                service.elapsedMillis.collect { elapsed ->
                    _state.value = _state.value.copy(elapsedMillis = elapsed)
                }
            }
        }
        val students = debateRepository.getStudents(sessionId)
        val speakers = buildSpeakers(session, students)
        _state.value = _state.value.copy(
            session = session,
            students = students,
            speakers = speakers,
            isLoading = false
        )
    }

    private fun buildSpeakers(session: DebateSession, students: List<com.debatefeedback.domain.model.Student>): List<SpeakerDisplay> {
        val composition = session.teamComposition ?: TeamComposition(prop = students.take(3).map { it.id }, opp = students.drop(3).take(3).map { it.id })
        val order = composition.speakerOrder(session.format)
        return if (order.isNotEmpty()) {
            order.map { slot ->
                val student = students.firstOrNull { it.id == slot.studentId }
                SpeakerDisplay(
                    studentId = slot.studentId,
                    speakerName = student?.name ?: slot.position,
                    position = slot.position
                )
            }
        } else {
            students.mapIndexed { index, student ->
                SpeakerDisplay(student.id, student.name, "Speaker ${index + 1}")
            }
        }
    }

    fun startRecording() {
        val session = _state.value.session ?: return
        val speaker = currentSpeaker() ?: return
        if (_state.value.isRecording) return
        val timer = timerService ?: return
        viewModelScope.launch {
            runCatching {
                recordingService.startRecording(session.id, speaker.speakerName, speaker.position)
            }.onSuccess { file ->
                timer.reset()
                timer.start()
                _state.value = _state.value.copy(isRecording = true, activeRecordingFile = file)
            }.onFailure { error ->
                _state.value = _state.value.copy(errorMessage = error.localizedMessage)
            }
        }
    }

    fun stopRecording() {
        if (!_state.value.isRecording) return
        val session = _state.value.session ?: return
        val speaker = currentSpeaker() ?: return
        val timer = timerService ?: return
        viewModelScope.launch {
            val result = recordingService.stopRecording()
            timer.stop()
            if (result == null) {
                _state.value = _state.value.copy(isRecording = false, activeRecordingFile = null)
                return@launch
            }
            val recording = SpeechRecording(
                speakerName = speaker.speakerName,
                speakerPosition = speaker.position,
                studentId = speaker.studentId,
                localFilePath = result.file.absolutePath,
                durationSeconds = result.durationSeconds,
                debateSessionId = session.id
            )
            debateRepository.addRecording(recording)
            uploadRecording(recording, result.file)
            _state.value = _state.value.copy(isRecording = false, activeRecordingFile = null)
            advanceSpeaker()
        }
    }

    fun cancelRecording() {
        viewModelScope.launch {
            recordingService.cancelRecording()
            timerService?.reset()
            _state.value = _state.value.copy(isRecording = false, activeRecordingFile = null)
        }
    }

    fun previousSpeaker() {
        if (_state.value.isRecording) return
        val newIndex = (_state.value.currentSpeakerIndex - 1).coerceAtLeast(0)
        _state.value = _state.value.copy(currentSpeakerIndex = newIndex)
        timerService?.reset()
    }

    fun nextSpeaker() {
        if (_state.value.isRecording) return
        val newIndex = (_state.value.currentSpeakerIndex + 1).coerceAtMost(_state.value.speakers.lastIndex)
        _state.value = _state.value.copy(currentSpeakerIndex = newIndex)
        timerService?.reset()
    }

    private fun advanceSpeaker() {
        val current = _state.value.currentSpeakerIndex
        if (current < _state.value.speakers.size - 1) {
            _state.value = _state.value.copy(currentSpeakerIndex = current + 1)
        }
        timerService?.reset()
    }

    fun dismissError() {
        _state.value = _state.value.copy(errorMessage = null)
    }

    private fun currentSpeaker(): SpeakerDisplay? = _state.value.speakers.getOrNull(_state.value.currentSpeakerIndex)

    private fun uploadRecording(recording: SpeechRecording, file: File) {
        viewModelScope.launch {
            val session = _state.value.session ?: return@launch
            val debateId = session.backendDebateId ?: session.id
            val metadata = mapOf(
                "speaker_name" to recording.speakerName,
                "speaker_position" to recording.speakerPosition,
                "duration_seconds" to recording.durationSeconds.toString(),
                "student_level" to session.studentLevel.name.lowercase(),
                "content_type" to "audio/m4a"
            )
            val uploadingRecording = recording.copy(uploadStatus = UploadStatus.UPLOADING, transcriptionStatus = ProcessingStatus.PROCESSING, feedbackStatus = ProcessingStatus.PENDING)
            debateRepository.updateRecording(uploadingRecording)
            _state.value = _state.value.copy(uploadProgress = _state.value.uploadProgress + (recording.id to 0.0))
            runCatching {
                uploadRepository.uploadSpeech(debateId, file, metadata) { progress ->
                    _state.value = _state.value.copy(uploadProgress = _state.value.uploadProgress + (recording.id to progress))
                }
            }.onSuccess { response ->
                val updated = uploadingRecording.copy(
                    uploadStatus = UploadStatus.UPLOADED,
                    speechId = response.speechId,
                    feedbackStatus = ProcessingStatus.PROCESSING,
                    transcriptionStatus = ProcessingStatus.PROCESSING
                )
                debateRepository.updateRecording(updated)
                pollSpeechStatus(updated)
            }.onFailure { error ->
                val failed = recording.copy(uploadStatus = UploadStatus.FAILED)
                debateRepository.updateRecording(failed)
                _state.value = _state.value.copy(errorMessage = error.localizedMessage)
            }
        }
    }

    private fun pollSpeechStatus(recording: SpeechRecording) {
        val speechId = recording.speechId ?: return
        viewModelScope.launch {
            var latest = recording
            repeat(60) {
                delay(Constants.API.FEEDBACK_POLL_SECONDS * 1000)
                val status = runCatching { debateRepository.fetchSpeechStatus(speechId) }
                    .getOrElse { error ->
                        _state.value = _state.value.copy(errorMessage = error.localizedMessage)
                        return@repeat
                    }
                val updated = applyStatus(latest, status)
                debateRepository.updateRecording(updated)
                latest = updated
                if (updated.feedbackStatus == ProcessingStatus.COMPLETE || updated.feedbackStatus == ProcessingStatus.FAILED) {
                    return@launch
                }
            }
        }
    }

    private fun applyStatus(recording: SpeechRecording, status: SpeechStatusResponse): SpeechRecording {
        val transcriptionStatus = ProcessingStatus.fromApi(status.transcriptionStatus)
        val feedbackStatus = ProcessingStatus.fromApi(status.feedbackStatus)
        return recording.copy(
            transcriptionStatus = transcriptionStatus,
            feedbackStatus = feedbackStatus,
            transcriptUrl = status.transcriptUrl ?: recording.transcriptUrl,
            feedbackUrl = status.googleDocUrl ?: recording.feedbackUrl,
            feedbackErrorMessage = status.feedbackError ?: recording.feedbackErrorMessage,
            transcriptionErrorMessage = status.transcriptionError ?: recording.transcriptionErrorMessage
        )
    }
}

data class TimerUiState(
    val session: DebateSession? = null,
    val students: List<com.debatefeedback.domain.model.Student> = emptyList(),
    val speakers: List<SpeakerDisplay> = emptyList(),
    val currentSpeakerIndex: Int = 0,
    val elapsedMillis: Long = 0L,
    val isRecording: Boolean = false,
    val isLoading: Boolean = true,
    val errorMessage: String? = null,
    val recordings: List<SpeechRecording> = emptyList(),
    val uploadProgress: Map<String, Double> = emptyMap(),
    val lastBellCount: Int? = null,
    val bellTimestamp: Long? = null,
    val activeRecordingFile: File? = null
)

data class SpeakerDisplay(
    val studentId: String?,
    val speakerName: String,
    val position: String
)
